### ISSUE TITLE
[pigeon] Consider not sending null class fields

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#105233]
Currently the native code generators for data class serialization always write all of their fields, even when null. This seems wasteful, since it means we're encoding and decoding string values and null values, and then sending them across the channel, rather than just omitting them. It looks like currently the Dart code is just doing lookups and checking against null, so currently I don't think there would be a behavioral difference if we omitted them instead.

If we change this we should change it across all generators, just to make sure we don't accidentally include subtle behavioral differences that could confuse us later.

@gaaclarke Is there a reason it was written this way? I don't see any reason to send explicit nulls, but I could certainly be missing something.

**Comment by gaaclarke** [comment]
Yea, the idea of the explicit null is just a safeguard against mixmatch endpoints or to expose bugs (ie is the sender forgetting to send the value or does it intend the value to be null).  I think if we do this we should also consider doing the handshake when creating the plugin to make sure that the schemas match on both endpoints.

**Comment by gaaclarke** [comment]
FWIW null encoding and decoding is the fastest code path, albeit not faster than it not being there.

**Comment by stuartmorgan** [comment]
Yes, it's the strings for the field names that are the non-trivial wasteful part. There are definitely cases in the Pigeon migration I'm doing now where for common cases 80+% of the payload and encoding will be strings of null fields.

**Comment by stuartmorgan** [comment]
> Yea, the idea of the explicit null is just a safeguard against mixmatch endpoints or to expose bugs (ie is the sender forgetting to send the value or does it intend the value to be null).

We're not actually doing this though (at least not on the Dart side) that I can see:

https://github.com/flutter/packages/blob/28ad56dfb3fe8c83d404bde440724388d2516dd7/packages/pigeon/lib/dart_generator.dart#L491-L498

There's no validation that the key is present; missing or explicitly null will both just be silently null there.

**Comment by stuartmorgan** [comment]
Note from discussion: we don't want to do this if we're going to move to arrays for class serialization, since we'd need the placeholders.

**Comment by stuartmorgan** [comment]
Closing as obsolete due to #111503
